### PR TITLE
Add Operator CRD

### PR DESCRIPTION
**Pull request by dantingl**
Operator should own a CRD, add NCP replicas as the property in the CRD

### CODE REVIEW

**Comment by dantingl** [comment]
> Do we need to list the CRD in the role binding (role.yaml) ?

Sure, will add it in role binding

File: pkg/controller/configmap/config.go [hunk 1]
```diff
@@ -208,9 +208,13 @@ func Render(configmap *corev1.ConfigMap) ([]*unstructured.Unstructured, error) {
 		}
 	}
 	if haEnabled {
-		renderData.Data[operatortypes.NcpReplicasKey] = operatortypes.NcpHaReplicas
+		renderData.Data[operatortypes.NcpReplicasKey] = ncpReplicas
 	} else {
 		renderData.Data[operatortypes.NcpReplicasKey] = 1
+		if ncpReplicas != 1 {
+			log.Info(fmt.Sprintf("Set nsx-ncp deployment replicas to 1 instead of %d as HA is disabled",
```
Comment (ggverma) [inline pkg/controller/configmap/config.go:215]: this looks like a warning log. too bad this logger doesn't have it. maybe we can use its verbosity feature if we stick to this

File: pkg/controller/configmap/configmap_controller.go [hunk 2]
```diff
@@ -101,17 +108,37 @@ func (r *ReconcileConfigMap) Reconcile(request reconcile.Request) (reconcile.Res
...
 	}
 
+	// Fetch ncp-install CRD instance
+	ncpInstallCrd := &operatorv1.NcpInstall{}
+	err := r.client.Get(context.TODO(), types.NamespacedName{Name: operatortypes.NcpInstallCRDName,
+		Namespace: operatortypes.OperatorNamespace}, ncpInstallCrd)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info(fmt.Sprintf("%s CRD is not found", operatortypes.NcpInstallCRDName))
+			r.status.SetDegraded(statusmanager.OperatorConfig, "NoNcpInstallCRD",
```
Comment (ggverma) [inline pkg/controller/configmap/configmap_controller.go:124]: `"NoNcpInstallCRD"` should be a const here https://github.com/vmware/nsx-container-plugin-operator/blob/0471eb4eb93e11570cca8a9212e4e9182c51aceb/pkg/controller/statusmanager/status_manager.go#L31

File: pkg/controller/configmap/configmap_controller.go [hunk 3]
```diff
@@ -101,17 +108,37 @@ func (r *ReconcileConfigMap) Reconcile(request reconcile.Request) (reconcile.Res
...
+				fmt.Sprintf("%s CRD is not found", operatortypes.NcpInstallCRDName))
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		r.status.SetDegraded(statusmanager.OperatorConfig, "NoNcpInstallCRD",
+			fmt.Sprintf("Failed to get operator CRD: %v", err))
+		return reconcile.Result{}, err
```
Comment (ggverma) [inline pkg/controller/configmap/configmap_controller.go:131]: not requeueing but mentioned in the comment

File: pkg/controller/configmap/configmap_controller.go [hunk 4]
```diff
@@ -101,17 +108,37 @@ func (r *ReconcileConfigMap) Reconcile(request reconcile.Request) (reconcile.Res
...
+		Namespace: operatortypes.OperatorNamespace}, ncpInstallCrd)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info(fmt.Sprintf("%s CRD is not found", operatortypes.NcpInstallCRDName))
+			r.status.SetDegraded(statusmanager.OperatorConfig, "NoNcpInstallCRD",
+				fmt.Sprintf("%s CRD is not found", operatortypes.NcpInstallCRDName))
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		r.status.SetDegraded(statusmanager.OperatorConfig, "NoNcpInstallCRD",
```
Comment (ggverma) [inline pkg/controller/configmap/configmap_controller.go:129]: `"NoNcpInstallCRD"` should be `"InvalidNcpInstallCRD"`? it's present and thus not the same error

File: pkg/controller/configmap/configmap_controller.go [hunk 5]
```diff
@@ -101,17 +108,37 @@ func (r *ReconcileConfigMap) Reconcile(request reconcile.Request) (reconcile.Res
...
+			fmt.Sprintf("Failed to get operator CRD: %v", err))
+		return reconcile.Result{}, err
+	}
+	ncpReplicas := ncpInstallCrd.Spec.NcpReplicas
```
Comment (ggverma) [inline pkg/controller/configmap/configmap_controller.go:133]: could this be empty or null?

File: pkg/controller/configmap/configmap_controller.go [hunk 6]
```diff
@@ -221,13 +248,13 @@ func (r *ReconcileConfigMap) Reconcile(request reconcile.Request) (reconcile.Res
 		return reconcile.Result{}, err
 	}
 	if !ncpNeedChange && !agentNeedChange {
-		// Check if NCP_IMAGE changes
-		ncpImageChanged, err := r.isNcpImageChanged()
+		// Check if NCP_IMAGE and nsx-ncp replicas change
```
Comment (ggverma) [inline pkg/controller/configmap/configmap_controller.go:251]: `Check if NCP_IMAGE and nsx-ncp replicas change` -> `Check if NCP_IMAGE or nsx-ncp replicas change` or just remove the comment

File: pkg/controller/configmap/configmap_controller.go (hunk 4 above)
Comment (ggverma) [inline pkg/controller/configmap/configmap_controller.go:129]: do we need to unset `"NoNcpInstallCRD"`?

File: pkg/controller/configmap/config.go (hunk 1 above)
Comment (dantingl) [reply pkg/controller/configmap/config.go:215]: Sure, I'm also thinking to use the log level in our code. How about I update all logs in a following patch.

File: pkg/controller/configmap/configmap_controller.go (hunk 2 above)
Comment (dantingl) [reply pkg/controller/configmap/configmap_controller.go:124]: It is in OperatorConfig status level, NoNcpInstallCRD is just the reason.

File: pkg/controller/configmap/configmap_controller.go (hunk 4 above)
Comment (dantingl) [reply pkg/controller/configmap/configmap_controller.go:129]: Sure

Comment (dantingl) [reply pkg/controller/configmap/configmap_controller.go:129]: It should be done in line 327

File: pkg/controller/configmap/configmap_controller.go (hunk 3 above)
Comment (dantingl) [reply pkg/controller/configmap/configmap_controller.go:131]: Looks like configmap and network CRD have the same issue :)

File: pkg/controller/configmap/configmap_controller.go (hunk 5 above)
Comment (dantingl) [reply pkg/controller/configmap/configmap_controller.go:133]: Based on my test, the default value is 0. So looks like I need to handle the unset case

File: pkg/controller/configmap/configmap_controller.go (hunk 6 above)
Comment (dantingl) [reply pkg/controller/configmap/configmap_controller.go:251]: Sure

File: pkg/controller/configmap/config.go (hunk 1 above)
Comment (ggverma) [reply pkg/controller/configmap/config.go:215]: sounds good

File: pkg/controller/configmap/configmap_controller.go (hunk 2 above)
Comment (ggverma) [reply pkg/controller/configmap/configmap_controller.go:124]: I see, thanks Danting! :)

File: pkg/controller/configmap/configmap_controller.go (hunk 5 above)
Comment (ggverma) [reply pkg/controller/configmap/configmap_controller.go:133]: oh yeah, integers in go are 0 by default

**Comment by dantingl** [comment]
> is this for K8s? Sorry, I am not sure I understand why we need this CRD

For the certification, operator should own a CRD and update status on the CRD.

**Pull request merged**